src/Servers/Kestrel/Core/src/Internal/Http2/Http2StreamOfT.cs

```diff
@@ -9,22 +13,81 @@ namespace Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http2
 {
     internal sealed class Http2Stream<TContext> : Http2Stream, IHostContextContainer<TContext> where TContext : notnull
     {
-        private readonly IHttpApplication<TContext> _application;
+        private readonly Task _executingTask;
+        private bool _disposed;
 
-        public Http2Stream(IHttpApplication<TContext> application, Http2StreamContext context) 
+        private readonly Channel<ReadResult> _requestQueue = Channel.CreateUnbounded<ReadResult>(new UnboundedChannelOptions
+        {
+            // We want to run continuations inline
+            SingleReader = true,
+            SingleWriter = true,
+            AllowSynchronousContinuations = true
+        });
+
+        public Http2Stream(IHttpApplication<TContext> application, Http2StreamContext context)
         {
             Initialize(context);
-            _application = application;
+
+            _executingTask = ProcessRequestsAsync(application);
         }
 
         public override void Execute()
         {
             KestrelEventSource.Log.RequestQueuedStop(this, AspNetCore.Http.HttpProtocol.Http2);
-            // REVIEW: Should we store this in a field for easy debugging?
-            _ = ProcessRequestsAsync(_application);
+
+            _requestQueue.Writer.TryWrite(new ReadResult(default, isCanceled: false, isCompleted: false));
         }
 
         // Pooled Host context
         TContext? IHostContextContainer<TContext>.HostContext { get; set; }
+
+        protected override bool BeginRead(out ValueTask<ReadResult> awaitable)
+        {
+            awaitable = _requestQueue.Reader.ReadAsync();
+            return true;
+        }
+
+        protected override bool TryParseRequest(ReadResult result, out bool endConnection)
+        {
+            if (result.IsCompleted)
+            {
+                endConnection = true;
+                return true;
+            }
+
+            return base.TryParseRequest(result, out endConnection);
+        }
+
+        protected override bool EndRequestProcessing()
+        {
+            base.EndRequestProcessing();
+            return _requestQueue.Reader.Completion.IsCompleted;
```

Comment:
```suggestion
            return base.EndRequestProcessing() || return _requestQueue.Reader.Completion.IsCompleted;
```

If base.EndRequestProcessing() returns false, that means `_keepAlive` is false. We should factor `_keepAlive` into our [CanReuse logic](https://github.com/dotnet/aspnetcore/blob/4c2168ca4845b4da8ccad2132a3256c9777ef1ad/src/Servers/Kestrel/Core/src/Internal/Http2/Http2Stream.cs) so we don't pool it after we stop the loop.

Reply:
This should fix the test failures.

Reply:
I didn't see that override EndRequestProcessing() in both Http2Stream *and* Http2StreamOfT until now. That means the proposed fix doesn't work as-is, but there's no reason this override cannot be merged into the Http2Stream override and do the right thing. I think everything but the ctor change can be moved to Http2Stream.